File: clrs/7/hoare_quicksort.py
#!/usr/bin/env python
from randata import getrandata

def hoare_partition(arr, low, high):
    pivot = arr[low]
    i = low - 1
    j = high + 1
    while True:
        i += 1
        j -= 1
        while arr[j] > pivot:
            j -= 1
        while arr[i] < pivot:
            i += 1
        if i < j:
            arr[i], arr[j] = arr[j], arr[i]
        else:
            return j

def quicksort(arr, low, high):
    if low < high:
        mid = hoare_partition(arr, low, high)
        quicksort(arr, low, mid)
        quicksort(arr, mid + 1, high)


if __name__ == '__main__':
    arr = getrandata(0)
    print(arr)
    quicksort(arr, 0, -1)
    print(arr)

File: clrs/7/hoare_quicksort.c
#include <stdio.h>
#include <stdlib.h>
#include <time.h>

#define SWAP(a, b) { int tmp = a; a = b; b = tmp; }
#define N 0

int hoare_partition(int *arr, int low, int high)
{
    int x = arr[low],
        i = low - 1,
        j = high + 1;
    while (1) {
        do {
            j--;
        } while (arr[j] > x);
        do {
            i++;
        } while (arr[i] < x);
        if (i < j) {
            SWAP(arr[i], arr[j]);
        } else {
            return j;
        }
    }
}

void hoare_quicksort(int *arr, int low, int high)
{
    if (low < high) {
        int mid = hoare_partition(arr, low, high);
        hoare_quicksort(arr, low, mid);
        hoare_quicksort(arr, mid + 1, high);
    }
}

int *get_randata(int n)
{
    int *buf, i;
    buf = (int *) calloc(n, sizeof(int));
    if (!buf) {
        return NULL;
    }
    srand((unsigned) time(NULL));
    for (i = 0; i < n; i++) {
        buf[i] = rand() % 100;
    }
    return buf;
}

void print_arr(int *arr, int n)
{
    int i;
    for (i = 0; i < n; i++) {
        printf("%d ", arr[i]);
    }
    printf("\n");
}

int main(void)
{
    int *arr;
    arr = get_randata(N);
    if (arr) {
        int arr2[] = {1, 1, 1, 1, 1};
        print_arr(arr, N);
        hoare_quicksort(arr, 0, N - 1);
        print_arr(arr, N);
        print_arr(arr2, 5);
        hoare_quicksort(arr2, 0, 4);
        print_arr(arr2, 5);
    }

    return 0;
}

